posts/2021-04-21-What I wish somebody told me when I was learning Haskell.md

```diff
@@ -0,0 +1,337 @@
+
+As you learn, you will realise that while succinct code takes longer to read per line, it is much faster overall - so you are saving yourself time by making it more succinct. Also, many libraries would use more succinct functional style, so getting fluent with it will pay back very quickly.
+
+### Do not worry about the glue code
+
+Most of us use computers having quite vague ideas about how they work. Yet when it comes to reading the code, we have an expectation that we would have an equal understanding of every line and every character in our code.
+
+With Haskell it helps to differentiate between your application domain code, that you want to be as readable as possible, and the "glue code", that simply connects different things.
+
+Such glue are, for example, various "lift" functions that allow to compose different Monad types. Next time you stumble on something that looks like "lift", you can skip looking up what exactly it is - focus on the types around it and the functions that do something meaningful to your application - if compiler is happy and your code passes tests, understanding every bit of glue code is unnecessary.
+
+What helps is making reusable functions with short names that minimize the cognitive overhead.
+
+For example, one of the pieces of glue I found useful was this function:
+
+```haskell
+liftError :: (MonadIO m, MonadError e' m) => (e -> e') -> ExceptT e IO a -> m a
+liftError f a = liftIO (first f <$> runExceptT a) >>= liftEither
+```
+
+While it may look scary inside, like a computer, you don't need to understand how it works to be able to use it effectively - it is enough to understand by looking at its type that it "lifts" an action expressed as `ExceptT e IO` monad to a more generic monad `(MonadIO m, MonadError e' m)`, transforming the error type at the same time - it is useful when composing different libraries or modules each using their own error type.
+
+An equivalent code in `do` notation might be easier to parse for the beginners:
+
+```haskell
+liftError :: (MonadIO m, MonadError e' m) => (e -> e') -> ExceptT e IO a -> m a
+liftError f a = do
+  b <- liftIO $ runExceptT a
+  let c = case b of -- this case is Bifunctor `first` method used above
+        Left e -> Left (f e)
+        Right x -> Right x
+  liftEither c
+```
+
+But the point is that you do not need to understand either code to effectively use this function, and you will see that short functions are easier to maintain.
+
+## Embrace purity
+
+### Functions do nothing
+
+In a functional language a function does nothing - it simply defines a transformation between types - but it does not perform this transformation until it is... It is tempting to say "called", but the functions are not "called" in functional programming languages in the same way - they are evaluated and combined in bigger transformations, until the whole program is constructed.
```

Comment:
... but the functions are not "called" in functional programming languages ~~in the same way~~ - ...

Or elaborate "in the same way as ...", but I think shorter is better - it's quite clear you're comparing functional with imperative